pkg/root/trusted_root_create.go

```diff
@@ -0,0 +1,378 @@
+	}
+
+	for _, tsaCertChain := range tsaCertChains {
+		tsaCA, err := certsToAuthority(tsaCertChain)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse TSA certificate chain: %w", err)
+		}
+		tr.timestampingAuthorities = append(tr.timestampingAuthorities, *tsaCA)
+	}
+
+	return tr, nil
+}
+
+func pubkeyToTransparencyLogInstance(keyBytes []byte, tm time.Time) (*TransparencyLog, string, error) {
+	logId := sha256.Sum256(keyBytes)
+	der, _ := pem.Decode(keyBytes)
+	key, keyDetails, err := getKeyWithDetails(der.Bytes)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return &TransparencyLog{
+		BaseURL:             "",
+		ID:                  logId[:],
+		ValidityPeriodStart: tm,
+		HashFunc:            crypto.SHA256, // we can't get this from the keyBytes, assume SHA256
+		PublicKey:           key,
+		SignatureHashFunc:   keyDetails,
+	}, hex.EncodeToString(logId[:]), nil
+}
+
+func getKeyWithDetails(key []byte) (crypto.PublicKey, crypto.Hash, error) {
+	var k any
+	var hashFunc crypto.Hash
+	var err1, err2 error
+
+	k, err1 = x509.ParsePKCS1PublicKey(key)
+	if err1 != nil {
+		k, err2 = x509.ParsePKIXPublicKey(key)
+		if err2 != nil {
+			return 0, 0, fmt.Errorf("can't parse public key with PKCS1 or PKIX: %w, %w", err1, err2)
+		}
+	}
+
+	switch v := k.(type) {
+	case *ecdsa.PublicKey:
+		switch v.Curve {
+		case elliptic.P256():
+			hashFunc = crypto.SHA256
+		case elliptic.P384():
+			hashFunc = crypto.SHA384
+		case elliptic.P521():
+			hashFunc = crypto.SHA512
+		default:
+			return 0, 0, fmt.Errorf("unsupported elliptic curve %T", v.Curve)
+		}
+	case *rsa.PublicKey:
+		switch v.Size() * 8 {
+		case 2048, 3072, 4096:
+			hashFunc = crypto.SHA256
+		default:
+			return 0, 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+		}
+	default:
+		return 0, 0, errors.New("unknown public key type")
+	}
+
+	return k, hashFunc, nil
+}
+
+func certsToAuthority(certChainPem []byte) (*CertificateAuthority, error) {
+	var cert *x509.Certificate
+	var err error
+	rest := certChainPem
+	certChain := []*x509.Certificate{}
+
+	// skip potential whitespace at end of file (8 is kinda random, but seems to work fine)
+	for len(rest) > 8 {
+		var derCert *pem.Block
+		derCert, rest = pem.Decode(rest)
+		if derCert == nil {
+			return nil, fmt.Errorf("input is left, but it is not a certificate: %+v", rest)
+		}
+		cert, err = x509.ParseCertificate(derCert.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse certificate: %w", err)
+		}
+		certChain = append(certChain, cert)
+	}
+	if len(certChain) == 0 {
+		return nil, fmt.Errorf("no certificates found in input")
+	}
+
+	ca := CertificateAuthority{}
+
+	for i, cert := range certChain {
+		switch {
+		case i == 0 && !cert.IsCA:
+			ca.Leaf = cert
+		case i < len(certChain)-1:
+			ca.Intermediates = append(ca.Intermediates, cert)
+		case i == len(certChain)-1:
+			ca.Root = cert
+		}
+	}
+
+	// TODO: should this rather be the innermost cert?
```

Comment:
That's a good point. I think it should be the most restrictive validity period of all of the CA certificates, which is likely the intermediate lowest in the chain.

Reply:
Sounds good, I can fix that.

Reply:
Done in https://github.com/sigstore/sigstore-go/pull/247/commits/b4094db25d0cf925e9dcdbabc63e23e08b0f37d8